lib/src/json_schema/json_schema.dart

```diff
@@ -630,9 +630,18 @@ class JsonSchema {
             }
           }
 
-          // If currentSchema is a ref, resolve ref recursively.
-          // Cycle detection happens at evaluation time for drafts 2019 and 2020.
-          if (currentSchema.schemaVersion < SchemaVersion.draft2019_09 && currentSchema.ref != null) {
+          // If currentSchema is a ref, we might want to resolve it recursively right now.
+          if (currentSchema.ref != null) {
+            // If we are at the end of the fragments to search and there are additional properties in the schema,
+            // continue with the current schema instead of resolving the ref.
+            if (i + 1 == fragments.length && currentSchema._schemaMap.length > 1) {
```

Comment:
I don't think length is an adequate check here. I think we should make a util to check for ref and no other schema-modifying properties. i.e. $ref + $schema or $ref + $id would be ok.